runtime/vm/VMAccess.cpp

```diff
@@ -261,18 +270,15 @@ acquireExclusiveVMAccess(J9VMThread * vmThread)
 				}
 #endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
 #if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
-				if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
-					VM_AtomicSupport::readBarrier(); // necessary?
-					if (currentThread->inNative) {
+				if (currentThread->inNative) {
 #if !defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
-						VM_VMAccess::setPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
+					VM_VMAccess::setPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
 #endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-					} else {
+				} else if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
 #if defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
-						VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
+					VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
 #endif /* J9VM_INTERP_TWO_PASS_EXCLUSIVE */
-						responsesExpected++;
-					}
+					responsesExpected++;
```

Comment:
Can this and the `#else` branch be merged into 1 copy?
```
#if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
				if (currentThread->inNative) {
#if !defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
					VM_VMAccess::setPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
#endif /* !J9VM_INTERP_TWO_PASS_EXCLUSIVE */
				} else 
#endif /* J9VM_INTERP_ATOMIC_FREE_JNI */
				if (currentThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
#if defined(J9VM_INTERP_TWO_PASS_EXCLUSIVE)
					VM_VMAccess::clearPublicFlags(currentThread, J9_PUBLIC_FLAGS_NOT_COUNTED_BY_EXCLUSIVE);
#endif /* J9VM_INTERP_TWO_PASS_EXCLUSIVE */
					responsesExpected++;
				}
```

Reply:
I'm intentionally keeping the old and new paths separate for now.  Once the new feature has been proven, the old code will be removed.